File: mobile/lib/models/game.dart
import 'match.dart';

class Game {
  final String id;
  final String name;
  final List<Match> competition;
  final DateTime startDate;
  final DateTime? endDate;
  final int maxLives;
  final bool isActive;
  final int currentWeek;
  final int totalWeeks;
  final int participantCount;
  final int maxParticipants;
  final DateTime createdAt;
  final DateTime updatedAt;

  Game({
    required this.id,
    required this.name,
    required this.competition,
    required this.startDate,
    this.endDate,
    required this.maxLives,
    required this.isActive,
    required this.currentWeek,
    required this.totalWeeks,
    required this.participantCount,
    required this.maxParticipants,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Game.fromJson(Map<String, dynamic> json) {
    return Game(
      id: json['_id'] ?? json['id'] ?? '',
      name: json['name'] ?? '',
      competition: (json['competition'] as List<dynamic>?)
          ?.map((match) => Match.fromJson(match))
          .toList() ?? [],
      startDate: json['startDate'] != null ? DateTime.parse(json['startDate']) : DateTime.now(),
      endDate: json['endDate'] != null ? DateTime.parse(json['endDate']) : null,
      maxLives: json['maxLives'] ?? 3,
      isActive: json['isActive'] ?? true,
      currentWeek: json['currentWeek'] ?? 1,
      totalWeeks: json['totalWeeks'] ?? 1,
      participantCount: json['participantCount'] ?? 0,
      maxParticipants: json['maxParticipants'] ?? 20,
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : DateTime.now(),
      updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'competition': competition.map((match) => match.toJson()).toList(),
      'startDate': startDate.toIso8601String(),
      'endDate': endDate?.toIso8601String(),
      'maxLives': maxLives,
      'isActive': isActive,
      'currentWeek': currentWeek,
      'totalWeeks': totalWeeks,
      'participantCount': participantCount,
      'maxParticipants': maxParticipants,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  List<Match> get currentWeekMatches {
    // Por simplicidad, asumimos 3 partidos por semana
    const matchesPerWeek = 3;
    final startIndex = (currentWeek - 1) * matchesPerWeek;
    final endIndex = startIndex + matchesPerWeek;
    
    if (startIndex >= competition.length) return [];
    
    return competition.sublist(
      startIndex, 
      endIndex > competition.length ? competition.length : endIndex
    );
  }

  bool get isFinished => currentWeek > totalWeeks;
}

File: backend/models/Game.ts
import mongoose, { Document, Schema } from 'mongoose';

export interface ITeam {
  name: string;
  flag: string;
}

export interface IMatch {
  matchId: string;
  home: ITeam;
  visitor: ITeam;
  date: Date;
  week: number; // Jornada a la que pertenece este partido
  result?: {
    homeScore: number;
    visitorScore: number;
    winner: 'home' | 'visitor' | 'draw';
  };
  isFinished: boolean;
  status: 'pending' | 'in_progress' | 'finished';
  bettingDeadline: Date;
  predictionsProcessed?: boolean;
}

export interface IWeek {
  weekNumber: number;
  name: string; // "Jornada 1", "Jornada 2", etc.
  startDate: Date;
  endDate: Date;
  matches: IMatch[];
  isActive: boolean;
  isCompleted: boolean;
}

export interface IGame extends Document {
  name: string;
  weeks: IWeek[]; // Nuevo: Array de jornadas
  competition: IMatch[]; // Mantener para compatibilidad, pero se llenará desde weeks
  startDate: Date;
  endDate?: Date;
  maxLives: number;
  isActive: boolean;
  currentWeek: number;
  totalWeeks: number;
  maxParticipants: number;
  participantCount: number;
  createdAt: Date;
  updatedAt: Date;
}

const TeamSchema = new Schema({
  name: { type: String, required: true, trim: true },
  flag: { type: String, required: true, trim: true }
});

const MatchSchema = new Schema({
  matchId: { type: String, required: true },
  home: { type: TeamSchema, required: true },
  visitor: { type: TeamSchema, required: true },
  date: { type: Date, required: true },
  week: { type: Number, required: true, min: 1 }, // Jornada
  result: {
    homeScore: { type: Number, min: 0 },
    visitorScore: { type: Number, min: 0 },
    winner: { type: String, enum: ['home', 'visitor', 'draw'] }
  },
  isFinished: { type: Boolean, default: false },
  status: { 
    type: String, 
    enum: ['pending', 'in_progress', 'finished'], 
    default: 'pending' 
  },
  bettingDeadline: { type: Date, required: true },
  predictionsProcessed: { type: Boolean, default: false }
});

const WeekSchema = new Schema({
  weekNumber: { type: Number, required: true, min: 1 },
  name: { type: String, required: true, trim: true },
  startDate: { type: Date, required: true },
  endDate: { type: Date, required: true },
  matches: [MatchSchema],
  isActive: { type: Boolean, default: false },
  isCompleted: { type: Boolean, default: false }
});

const GameSchema: Schema = new mongoose.Schema({
  name: { 
    type: String, 
    required: true, 
    trim: true,
    maxlength: 100
  },
  weeks: [WeekSchema], // Nuevo: Array de jornadas
  competition: [MatchSchema], // Mantener para compatibilidad
  startDate: { 
    type: Date, 
    required: true 
  },
  endDate: { 
    type: Date 
  },
  maxLives: { 
    type: Number, 
    default: 3,
    min: 1,
    max: 10
  },
  isActive: { 
    type: Boolean, 
    default: true 
  },
  currentWeek: { 
    type: Number, 
    default: 1,
    min: 1
  },
  totalWeeks: { 
    type: Number, 
    required: true,
    min: 1
  },
  maxParticipants: { 
    type: Number, 
    required: true, 
    min: 1, 
    default: 20 
  },
  participantCount: { 
    type: Number, 
    default: 0 
  }
}, {
  timestamps: true
});

export default mongoose.model<IGame>('Game', GameSchema);
